### PR TITLE
Add a time derivative auxkernel

### PR DESCRIPTION
**Pull request by GiudGiud**
Tentatively in support of https://github.com/idaholab/moose/discussions/20850

But honestly I was thinking I wanted to have this for a long time

If the FE functor implementations catch up we can remove the variable argument. For now we just have plenty of disclaimers in the docs.

### CODE REVIEW

File: test/tests/auxkernels/time_derivative_aux/tests [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      expect_err = 'Unable to couple time derivative of an auxiliary variable into the auxiliary system.'
+      detail = 'a time derivative is requested from an auxiliary variable for which the coupled interface does not support the time derivative'
+    []
+    [not_implemented_function]
+      type = RunException
+      input = 'test.i'
+      cli_args = 'AuxKernels/function_derivative_quadrature_point/functor=no_known_derivative'
+      expect_err = 'timeDerivative method not implemented'
+      detail = 'a time derivative is requested from a function for which the functor time derivative not implemented'
+    []
```
Comment (GiudGiud) [inline test/tests/auxkernels/time_derivative_aux/tests:40]: umm that doesnt really test the right thing. I ll remove that one

**Comment by GiudGiud** [comment]
@dschwen there's a name conflict with marmot for this object. I dont have access to marmot right now. How hard would the patch be? Should I ask for access?

File: framework/src/auxkernels/TimeDerivativeAux.C [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+TimeDerivativeAuxTempl<is_ad>::TimeDerivativeAuxTempl(const InputParameters & parameters)
+  : AuxKernel(parameters),
+    _v_dot(isParamValid("v") ? &coupledDot("v") : nullptr),
+    _functor(isParamValid("functor") ? &getFunctor<GenericReal<is_ad>>("functor") : nullptr),
+    _factor(getFunctor<GenericReal<is_ad>>("factor")),
+    _use_qp_arg(dynamic_cast<MooseVariableFE<Real> *>(&_var))
+{
+  if ((!_functor && !_v_dot) || (_functor && _v_dot))
+    mooseError("A variable or a functor parameter should be provided to the "
+               "TimeDerivativeAuxTempl<is_ad>");
```
Comment (lindsayad) [inline framework/src/auxkernels/TimeDerivativeAux.C:40]: Maybe add `name()` somewhere in here

File: framework/src/auxkernels/TimeDerivativeAux.C [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+      return raw_value(_factor(qp_arg)) * (*_v_dot)[_qp];
+    else
+      return raw_value(_factor(qp_arg)) * raw_value((*_functor).dot(qp_arg));
+  }
+  else
+  {
+    const auto elem_arg = makeElemArg(_current_elem);
+
+    if (_v_dot)
+      return raw_value(_factor(elem_arg)) * (*_v_dot)[_qp];
```
Comment (lindsayad) [inline framework/src/auxkernels/TimeDerivativeAux.C:71]: This seems a little awkward since we are in the `!_use_qp_arg` branch. Would it work to call `v_var.dot(elem_arg)` instead?

File: framework/src/auxkernels/TimeDerivativeAux.C (hunk 2 above)
Comment (loganharbour) [reply framework/src/auxkernels/TimeDerivativeAux.C:40]: None of this is needed - the type and name are output with `MooseObject::mooseError()`

File: framework/src/auxkernels/TimeDerivativeAux.C (hunk 3 above)
Comment (GiudGiud) [reply framework/src/auxkernels/TimeDerivativeAux.C:71]: I dont think the dot is implemented for FE variables rn.
I could try to do that though, likely isnt too hard

**Comment by dschwen** [comment]
> @dschwen there's a name conflict with marmot for this object. I dont have access to marmot right now. How hard would the patch be (eg do you have something better in marmot? or essentially the same save for functor capability)? Should I ask for access?

I'll take a look. We could probably delete the Marmot object, or migrate it to moose, making you PR superfluous :-D

**Comment by dschwen** [comment]
We can remove the Marmot object. Your parameter `v` here is called `u` in the Marmot version, and the Marmot AuxKernel is absolutely bare bones (just `return _u_dot[_qp];`). So we can delete that ant the accompanying  test file, which contains the sole use of that object.

File: framework/src/auxkernels/TimeDerivativeAux.C (hunk 3 above)
Comment (lindsayad) [reply framework/src/auxkernels/TimeDerivativeAux.C:71]: `_use_qp_arg` is always true for FE variables right? Or I guess that you're imagining the aux var could be FV and the functor could be an FE variable? I actually have `evaluateDot` with qp arguments implemented for all variables:
```c++
template <typename OutputType>
typename MooseVariableField<OutputType>::DotType
MooseVariableField<OutputType>::evaluateDot(const ElemQpArg & elem_qp,
                                            const unsigned int state) const
{
  mooseAssert(_time_integrator && _time_integrator->dt(),
              "A time derivative is being requested but we do not have a time integrator so we'll "
              "have no idea how to compute it");
  evaluateOnElement(elem_qp, state);
  const auto qp = std::get<1>(elem_qp);
  mooseAssert(qp < _current_elem_qp_functor_dot.size(),
              "The requested " << qp << " is outside our dot size");
  return _current_elem_qp_functor_dot[qp];
}

template <typename OutputType>
void
MooseVariableField<OutputType>::evaluateOnElementSide(const ElemSideQpArg & elem_side_qp,
                                                      const unsigned int state) const
{
```
The issue is that the element-based arguments aren't implemented for FE

Comment (GiudGiud) [reply framework/src/auxkernels/TimeDerivativeAux.C:71]: well except we're already in the "not a FE variable" part of the loop. So yeah we can use v_var, it ll be a finite volume variable.

Comment (lindsayad) [reply framework/src/auxkernels/TimeDerivativeAux.C:71]: We're in the `_var` is not an FE variable part of the loop. We have not done any checks on `v_var`

Comment (lindsayad) [reply framework/src/auxkernels/TimeDerivativeAux.C:71]: Ideally we don't see mixing of `elem_arg` and `_qp` on the same line

**Comment by dschwen** [comment]
Marmot MR is up. I'll let you know when it's merged.

**Comment by dschwen** [comment]
Merged

File: framework/src/auxkernels/TimeDerivativeAux.C (hunk 3 above)
Comment (GiudGiud) [reply framework/src/auxkernels/TimeDerivativeAux.C]: should be unmixed now.
Would be someone using a FE variable time derivative to fill a FV variable, wont happen for a while haha

I added a rider commit with a warning if you dont mind.

Comment (lindsayad) [reply framework/src/auxkernels/TimeDerivativeAux.C]: When the aux system gets involved, I feel like an accidental mixing of FE and FV is more likely since a user may just copy stuff like
```c++
[AuxVariables]
  [aux]
    family = MONOMIAL
    order = CONSTANT
  []
[]
```
But yes, I agree that someone will not be doing this *on purpose* for quite a while.

However, I think your new error in `Coupleable.h` helps highlight that actually the *only* way we can get to this line of code is for the aux variable to be FV and the coupled variable to be FE ... right? Because `coupledDot` will only return without erroring if the coupled variable is an FE variable. This line is currently uncovered in testing; actually the whole FV branch is uncovered because the `test_fv.i` has its aux variable as FE (some more of that mixing :smile: ).

My suggestion would be that we do not populate `_v_dot` in the initialization list (well populate it, but with `nullptr`), and instead only call `coupledDot` if `_v_var` dynamic casts to `MooseVariableFE<Real>` (or `MooseVariable` for short) and `_use_qp_arg`. Then in the FV branch of `computeValue`, change `if (_v_dot)` to `if (_v_var)`.

This change would allow a user, in an all FV simulation, to get the results they want, instead of an error, if they set the parameter `v` instead of `functor`.

Comment (GiudGiud) [reply framework/src/auxkernels/TimeDerivativeAux.C]: Well I dont know what the solution is here. I added some warnings, for the `v` case. I would much rather we delete that parameter tbh, but that would make an object that is unfit for what the discussions user I m trying to help is doing. 

I reworked this a little to add warnings with the implicit conversions that we are doing. This is of sort of limited, to the v parameter case, because when we input a functor:
- we dont know if it s a fv or fe variable (we could with some more code I guess, getFieldVar("functor"), catch it when it s not even variable etc)
- a power user could be trying to do these conversions. This will ultimately be part of the purpose of the functor system, enough flexibility for both FE and FV
 
so to summarize, I dont catch this with a warning:
```
[AuxVariables]
  [aux]
    family = MONOMIAL
    order = CONSTANT
  []
[]

[AuxKernels]
  [deriv]
    type = TimeDerivativeAux
    variable = aux
    functor = some_fv_variable
  []
[]
```
OR this
```
[AuxVariables]
  [aux_fv]
    family = MONOMIAL
    order = CONSTANT
    fv = true
  []
[]

[AuxKernels]
  [deriv]
    type = TimeDerivativeAux
    variable = aux_fv
    functor = some_fe_variable
  []
[]
```

File: framework/src/auxkernels/TimeDerivativeAux.C [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    if (_v_dot)
+      return raw_value(_factor(qp_arg)) * (*_v_dot)[_qp];
+    else
+      return raw_value(_factor(qp_arg)) * raw_value((*_functor).dot(qp_arg));
+  }
+  else
+  {
+    const auto elem_arg = makeElemArg(_current_elem);
+
+    if (_v_dot)
```
Comment (lindsayad) [inline framework/src/auxkernels/TimeDerivativeAux.C:82]:
```suggestion
    if (_v_var)
```

File: framework/src/auxkernels/TimeDerivativeAux.C (hunk 3 above)
Comment (GiudGiud) [reply framework/src/auxkernels/TimeDerivativeAux.C]: I think this extensive checking on FE-FV conversion will have to move up to AuxKernel.C, it s getting too big

Comment (lindsayad) [reply framework/src/auxkernels/TimeDerivativeAux.C]: >  I would much rather we delete that parameter tbh, but that would make an object that is unfit for what the discussions user I m trying to help is doing.

Why is that? It seems like all the dancing that is being done in here would be removed if we just did a functor evaluation with a quadrature point based argument. That should work in all cases because the qp-based functor evaluations are implemented in `MooseVariableField`. I thought maybe you wanted to incorporate `coupledDot` for speed reasons?

Anyway your most recent changes looked fine to me, other than just changing that last `if (_v_dot)` comparison to `if (_v_var)`

Comment (GiudGiud) [reply framework/src/auxkernels/TimeDerivativeAux.C]: not really. I was getting an error on INITIAL, thought it was a general issue with how I did things and moved to coupledDot because that was working and is the vetted way of doing this. But qp functor will do. 

My most recent changes are all gone. And so is v_dot :O

**Review by lindsayad — DISMISSED** [review]
I remembered that there is one potential issue ... if you are attempting to take the `.dot` of an auxiliary variable whose value was computed in a dependency aux kernel, the functor `.dot` evaluation will not have up-to-date information regarding that variable's entries in the solution vector. This is covered in https://github.com/idaholab/moose/issues/19396

**Comment by GiudGiud** [comment]
well if you did that with the coupled interface there was an error message so either way derivatives of auxvariables arent super supported

**Comment by GiudGiud** [comment]
Maybe we should add a warning

**Comment by GiudGiud** [comment]
need new approval, one downside of functors is needs global AD indexing for qp argument of FE variables.
nbd though since that s the standard rn

**Review by lindsayad — APPROVED** [review]

**Comment by GiudGiud** [comment]
Test failure is unrelated (libtorch).
